### PR TITLE
Fix MySQL Binlog XA Transaction

### PR DESCRIPTION
**Pull request by zhangdove**
Hi! 
Thranks For Maxwell Team!
There is a PR that Fix XA Transaction! 
[https://github.com/zendesk/maxwell/blob/master/src/test/resources/ddl/5.7/xa-statements](xa-statements)
hope this would be helpful!

my maxwell version is 1.22.1!

### CODE REVIEW

File: src/main/java/com/zendesk/maxwell/replication/BinlogConnectorReplicator.java [hunk 1]
```diff
@@ -498,6 +500,10 @@ private RowMapBuffer getTransactionRows(BinlogConnectorEvent beginEvent) throws
 					String sql = qe.getSql();
 					String upperCaseSql = sql.toUpperCase();
 
+					if (sql.startsWith(BinlogConnectorEvent.XAEND)) {
+						LOGGER.info("XA END");
+						return buffer;
+					}
```
Comment (osheroff) [inline src/main/java/com/zendesk/maxwell/replication/BinlogConnectorReplicator.java]: please move this into `.isCommitEvent()`, and then let the logic on line 458 trigger.

File: src/main/java/com/zendesk/maxwell/replication/BinlogConnectorReplicator.java [hunk 2]
```diff
@@ -606,6 +612,27 @@ public RowMap getRow() throws Exception {
 					} else {
 						processQueryEvent(event);
 					}
+					if (sql.startsWith(BinlogConnectorEvent.XASTART) || sql.startsWith(BinlogConnectorEvent.XABEGIN)) {
```
Comment (osheroff) [inline src/main/java/com/zendesk/maxwell/replication/BinlogConnectorReplicator.java]: let's move all of these `if` statements into the above `if` block.

File: src/main/java/com/zendesk/maxwell/replication/BinlogConnectorReplicator.java [hunk 3]
```diff
@@ -65,6 +66,7 @@
 	private final BootstrapController bootstrapper;
 	private final AbstractProducer producer;
 	private RowMapBuffer rowBuffer;
+	private ConcurrentHashMap<String, RowMapBuffer> rowBufferHashMapXA = new ConcurrentHashMap();
```
Comment (osheroff) [inline src/main/java/com/zendesk/maxwell/replication/BinlogConnectorReplicator.java]: only one thread is ever mutating this, no need for `ConcurrentHashMap` -- just use a regular `HashMap`

File: src/main/java/com/zendesk/maxwell/replication/BinlogConnectorReplicator.java [hunk 4]
```diff
@@ -623,10 +650,22 @@ public RowMap getRow() throws Exception {
 		}
 	}
 
+	private String getXid(String sql) {
```
Comment (osheroff) [inline src/main/java/com/zendesk/maxwell/replication/BinlogConnectorReplicator.java]: `Xid` generally refers to something else in maxwell, maybe we call this `XAid` or something?

File: src/main/java/com/zendesk/maxwell/replication/BinlogConnectorReplicator.java [hunk 5]
```diff
@@ -606,6 +612,27 @@ public RowMap getRow() throws Exception {
...
+					if (sql.startsWith(BinlogConnectorEvent.XASTART) || sql.startsWith(BinlogConnectorEvent.XABEGIN)) {
+						RowMapBuffer rowMapBufferXA = getTransactionRows(event);
+						String xid = getXid(sql);
+						rowBufferHashMapXA.put(xid, rowMapBufferXA);
+						LOGGER.info(BinlogConnectorEvent.XASTART + " rowBufferHashMapXA.size(): " + rowBufferHashMapXA.size() + " xid " + xid);
+					}
+					if (sql.startsWith(BinlogConnectorEvent.XACOMMIT)) {
+						String xid = getXid(sql);
+						rowBuffer = rowBufferHashMapXA.remove(xid);
+						updateStatus(rowBuffer);
```
Comment (osheroff) [inline src/main/java/com/zendesk/maxwell/replication/BinlogConnectorReplicator.java]: ok, full stop.  Can't there be multiple `XA COMMIT` events per `XA BEGIN / XA END` event?  If there can be, we'd be throwing away any previous XA transactions....

File: src/main/java/com/zendesk/maxwell/replication/BinlogConnectorReplicator.java (hunk 1 above)
Comment (zhangdove) [reply src/main/java/com/zendesk/maxwell/replication/BinlogConnectorReplicator.java]: OK!I think is a greater idea!

File: src/main/java/com/zendesk/maxwell/replication/BinlogConnectorReplicator.java (hunk 2 above)
Comment (zhangdove) [reply src/main/java/com/zendesk/maxwell/replication/BinlogConnectorReplicator.java]: OK!The method of processQueryEvent is deal with some DDL event.
I will move all of these if statements into the above if block.

File: src/main/java/com/zendesk/maxwell/replication/BinlogConnectorReplicator.java (hunk 4 above)
Comment (zhangdove) [reply src/main/java/com/zendesk/maxwell/replication/BinlogConnectorReplicator.java]: em....As your wish!

File: src/main/java/com/zendesk/maxwell/replication/BinlogConnectorReplicator.java (hunk 5 above)
Comment (zhangdove) [reply src/main/java/com/zendesk/maxwell/replication/BinlogConnectorReplicator.java]: Yes!We only deal with XA COMMIT event  and XA ROLLBACK event!

Here are some online system logs

> 19/09/22 18:12:00 INFO replication.BinlogConnectorReplicator: XA START rowBufferHashMapXA.size(): 1 xid 31302e32372e3231352e3230302e746d313536393134373132303130323238323436
19/09/22 18:12:00 INFO replication.BinlogConnectorReplicator: XA COMMIT rowBufferHashMapXA.size(): 0 xid 31302e32372e3231352e3230302e746d313536393134373132303130323238323436
19/09/22 18:12:00 INFO replication.BinlogConnectorReplicator: XA END
19/09/22 18:12:00 INFO replication.BinlogConnectorReplicator: XA START rowBufferHashMapXA.size(): 1 xid 31302e32372e3231352e3230302e746d313536393134373132303230323238323437
19/09/22 18:12:00 INFO replication.BinlogConnectorReplicator: XA END
19/09/22 18:12:00 INFO replication.BinlogConnectorReplicator: XA START rowBufferHashMapXA.size(): 2 xid 31302e32372e392e322e746d313536393134373132303231383233303133
19/09/22 18:12:00 INFO replication.BinlogConnectorReplicator: XA COMMIT rowBufferHashMapXA.size(): 1 xid 31302e32372e3231352e3230302e746d313536393134373132303230323238323437
19/09/22 18:12:00 INFO replication.BinlogConnectorReplicator: XA COMMIT rowBufferHashMapXA.size(): 0 xid 31302e32372e392e322e746d313536393134373132303231383233303133
19/09/22 18:12:01 INFO replication.BinlogConnectorReplicator: XA END
19/09/22 18:12:01 INFO replication.BinlogConnectorReplicator: XA START rowBufferHashMapXA.size(): 1 xid 31302e32372e392e322e746d313536393134373132313234333233303136
19/09/22 18:12:01 INFO replication.BinlogConnectorReplicator: XA COMMIT rowBufferHashMapXA.size(): 0 xid 31302e32372e392e322e746d313536393134373132313234333233303136

File: src/main/java/com/zendesk/maxwell/replication/BinlogConnectorReplicator.java (hunk 1 above)
Comment (zhangdove) [reply src/main/java/com/zendesk/maxwell/replication/BinlogConnectorReplicator.java]: Now! I think _XA END_ event is not belong to _isCommitEvent()_.
only _XA COMMIT_ event and _XA ROLLBACK_ belong to.

**Comment by zhangdove** [comment]
new  [PR](https://github.com/zendesk/maxwell/pull/1358)

**Comment by lix09** [comment]
> new [PR](https://github.com/zendesk/maxwell/pull/1358) about XA Transaction!



> new [PR](https://github.com/zendesk/maxwell/pull/1358) about XA Transaction!
Program parsing slows down and data is lost

**Pull request closed without merging**